File: tessoku-book/a12/main.c
#include <stdio.h>
#include <stdbool.h>

bool	is_lte_middle(long int *A, long int mid, long int N, long int K)
{
	// printf("mid=%ld\n", mid);
	long int sum = 0;
	for (int i = 0; i < N; i++)
	{
		sum += mid / A[i];
	}
	
	if (sum >= K)
		return (true);
	else
		return (false);
}

int	main(void)
{
	long int	N;
	long int	K;
	
	scanf("%ld", &N);
	long int A[N];

	scanf("%ld", &K);
	// printf("%ld %ld\n", N, K);

	for (int i = 0; i < N; i++)
	{
		scanf("%ld", &A[i]);
		// printf("%d\n", A[i]);
	}

	long int L = 1, R = 100000000;
	while (L < R)
	{
		long int M = (L + R) / 2;
		if (is_lte_middle(A, M, N, K))
			R = M;
		else
			L = M + 1;
	}
	printf("%ld\n", L);
	
	return (0);
}

File: tessoku-book/a12/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

func isLteMiddle(a []int, n, k, middle int) bool {
	sum := 0
	for i := 0; i < n; i++ {
		sum += middle / a[i]
	}
	if sum >= k {
		return true
	} else {
		return false
	}
}

func main() {
	sc.Buffer(make([]byte, 128), 500000)
	sc.Split(bufio.ScanWords)
	n := scanInt()
	k := scanInt()
	// fmt.Printf("%d %d\n", n, k)

	a := make([]int, n)
	for i := 0; i < n; i++ {
		a[i] = scanInt()
	}

	left := 1
	right := int(math.Pow10(10))
	for left < right {
		middle := (left + right) / 2
		if isLteMiddle(a, n, k, middle) {
			right = middle
		} else {
			left = middle + 1
		}
	}
	fmt.Printf("%d\n", left)
}

var sc = bufio.NewScanner(os.Stdin)

func scanInt() int {
	sc.Scan()
	return atoi(sc.Text())
}
func atoi(nStr string) int {
	i, err := strconv.Atoi(nStr)
	if err != nil {
		panic(err)
	}
	return i
}
func scanText() string {
	sc.Scan()
	return sc.Text()
}
